repose-aggregator/core/repose-core/src/main/scala/org/openrepose/powerfilter/ReposeFilter.scala

```diff
@@ -150,54 +148,57 @@ class ReposeFilter @Inject()(@Value(ReposeSpringProperties.NODE.NODE_ID) nodeId:
             optMetricRegistry,
             tracer
           ).doFilter(wrappedRequest, wrappedResponse)
-        })
 
-        doFilterChain match {
-          case Success(_) =>
-            logger.trace("{} -- Successfully processed request", nodeId)
-          case Failure(use: URISyntaxException) =>
+          logger.trace("{} -- Successfully processed request", nodeId)
+        }
+
+        // Set the Via header
+        if (wrappedResponse.isCommitted) {
+          wrappedResponse.uncommit()
+        }
+        responseHeaderService.setVia(wrappedRequest, wrappedResponse)
+
+        // Handle Throwables that arose while processing
+        val recoveredFilterChain = doFilterChain.recover {
+          case use: URISyntaxException =>
             logger.debug(s"$nodeId -- Invalid URI requested: ${wrappedRequest.getRequestURI}", use)
-            wrappedResponse.sendError(SC_BAD_REQUEST, "Error processing request", true)
-          case Failure(e: Exception) =>
+            wrappedResponse.sendError(SC_BAD_REQUEST, "Error processing request")
+          case e: Exception =>
             logger.error(s"$nodeId -- Issue encountered while processing filter chain.", e)
-            wrappedResponse.sendError(SC_BAD_GATEWAY, "Error processing request", true)
-          case Failure(t: Throwable) =>
+            wrappedResponse.sendError(SC_BAD_GATEWAY, "Error processing request")
+          case t: Throwable =>
             logger.error(s"$nodeId -- Error encountered while processing filter chain.", t)
-            wrappedResponse.sendError(SC_BAD_GATEWAY, "Error processing request", true)
-            setViaClearCommitAndClose(wrappedRequest, wrappedResponse, scope, startTime)
+            wrappedResponse.sendError(SC_BAD_GATEWAY, "Error processing request")
             throw t
         }
-        setViaClearCommitAndClose(wrappedRequest, wrappedResponse, scope, startTime)
-        logger.trace("ReposeFilter returning response...")
+
+        // Commit the response and record metrics for the request and response
+        wrappedResponse.commitToResponse()
+        closeSpan(wrappedResponse, scope)
+        optMetricRegistry.foreach { mr =>
+          markResponseCodeHelper(
+            mr,
+            wrappedResponse.getStatus,
+            System.currentTimeMillis - startTime,
+            logger.underlying)
+        }
+
+        // Either log the final message
+        // Or rethrow any Throwables that arose while processing and were not handled
+        // Always clear out the logging context
+        doFilterChain.foreach(_ => logger.trace("ReposeFilter returning response..."))
+        MDC.clear()
+        doFilterChain.get
```

Comment:
This should be `recoveredFilterChain`, as should the log message above.

Reply:
Updated.